airflow-core/src/airflow/api_fastapi/core_api/services/public/task_instances.py

```diff
@@ -0,0 +1,285 @@
+    dag_bag: DagBagDep,
+    body: PatchTaskInstanceBody,
+    session: SessionDep,
+    map_index: int | None = -1,
+    update_mask: list[str] | None = Query(None),
+) -> tuple[DAG, list[TI], dict]:
+    dag = dag_bag.get_dag(dag_id)
+    if not dag:
+        raise HTTPException(status.HTTP_404_NOT_FOUND, f"DAG {dag_id} not found")
+
+    if not dag.has_task(task_id):
+        raise HTTPException(status.HTTP_404_NOT_FOUND, f"Task '{task_id}' not found in DAG '{dag_id}'")
+
+    query = (
+        select(TI)
+        .where(TI.dag_id == dag_id, TI.run_id == dag_run_id, TI.task_id == task_id)
+        .join(TI.dag_run)
+        .options(joinedload(TI.rendered_task_instance_fields))
+    )
+    if map_index is not None:
+        query = query.where(TI.map_index == map_index)
+
+    tis = session.scalars(query).all()
+
+    err_msg_404 = (
+        f"The Task Instance with dag_id: `{dag_id}`, run_id: `{dag_run_id}`, task_id: `{task_id}` and map_index: `{map_index}` was not found.",
+    )
+    if len(tis) == 0:
+        raise HTTPException(status.HTTP_404_NOT_FOUND, err_msg_404)
+
+    fields_to_update = body.model_fields_set
+    if update_mask:
+        fields_to_update = fields_to_update.intersection(update_mask)
+    else:
+        try:
+            PatchTaskInstanceBody.model_validate(body)
+        except ValidationError as e:
+            raise RequestValidationError(errors=e.errors())
+
+    return dag, list(tis), body.model_dump(include=fields_to_update, by_alias=True)
+
+
+class BulkTaskInstanceService(BulkService[BulkTaskInstanceBody]):
+    """Service for handling bulk operations on task instances."""
+
+    def __init__(
+        self,
+        session: Session,
+        request: BulkBody[BulkTaskInstanceBody],
+        dag_id: str,
+        dag_run_id: str,
+        dag_bag: DagBagDep,
+        user: GetUserDep,
+    ):
+        super().__init__(session, request)
+        self.dag_id = dag_id
+        self.dag_run_id = dag_run_id
+        self.dag_bag = dag_bag
+        self.user = user
+
+    def categorize_task_instances(
+        self, task_ids: set
+    ) -> tuple[dict[tuple[str, int], TI], set[str], set[str]]:
+        """
+        Categorize the given task_ids into matched_task_ids and not_found_task_ids based on existing task_ids.
+
+        :param task_ids: set of task_ids
+        :return: tuple of (task_instances_map, matched_task_ids, not_found_task_ids)
+        """
+        query = select(TI).where(
+            TI.dag_id == self.dag_id, TI.run_id == self.dag_run_id, TI.task_id.in_(task_ids)
+        )
+        task_instances = self.session.scalars(query).all()
+        task_instances_map = {
+            (ti.task_id, ti.map_index if ti.map_index is not None else -1): ti for ti in task_instances
+        }
+        matched_task_ids = {task_id for (task_id, _) in task_instances_map.keys()}
+        not_found_task_ids = task_ids - matched_task_ids
+        return task_instances_map, matched_task_ids, not_found_task_ids
```

Comment:
I think `matched_task_ids` and `not_found_task_ids` should be tuples of `(task_id, map_index)` too.  This will later on simplify the implementation because we have two layers to check for existing tasks, the first layer is `task_id` only, then we do another check on the map_index too. We can do all of that in `categorize_task_instances` I believe.
```suggestion
        return task_instances_map, matched_task_ids, not_found_task_ids
```

Reply:
Sure, I've refactored it to a more simple version.